### PR TITLE
Refactor timer widget

### LINKED ISSUE #111: [FEATURE] Add timer to pilot gui
### Is your feature request related to a problem? Please describe

The timer widget tracks the time internally, so multiple timer widgets will not be synchronized

### Describe the solution you'd like

Abstract the actual timekeeping to a ros node that runs on the operator computer and modify the timer widget to communicate with it.

### PR DESCRIPTION
**Pull request by NoahMollerstuen**
# Pull Request

## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Abstract the timer widget into a separate node on the timer laptop communicating over ros with a simpler widget. This allows the global timer to be displayed on multiple places across both guis.

## Related Tickets & Documents
Closes #111

### CODE REVIEW

File: src/surface/gui/gui/auxiliary_nodes/timer.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import typing as t
```
Comment (benjaminwp18) [inline src/surface/gui/gui/auxiliary_nodes/timer.py:1]: Don't alias this

File: src/rov_msgs/msg/MissionTimer.msg
Comment (benjaminwp18) [inline src/rov_msgs/msg/MissionTimer.msg:1]: Rename this file to be less ambiguous with the actual mission timer. `MissionTimerTick`? `MissionTimerBroadcast`?

File: src/rov_msgs/srv/MissionTimerSetRunning.srv
Comment (benjaminwp18) [inline src/rov_msgs/srv/MissionTimerSetRunning.srv:1]: Consider removing this service in favor of an optional set_time field in `MissionTimerSetTime.srv`

File: src/surface/gui/gui/auxiliary_nodes/timer.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    def set_running_callback(self, request: MissionTimerSetRunning.Request,
+                             response: MissionTimerSetRunning.Response
+                             ) -> MissionTimerSetRunning.Response:
+        timestamp = self.get_clock().now()
+        if self.is_running and self.lastTimestamp is not None:
+            self.time_left = max(
+                self.lastTimestamp + self.time_left - timestamp,
+                Duration(seconds=0)
+            )
```
Comment (benjaminwp18) [inline src/surface/gui/gui/auxiliary_nodes/timer.py:77]: To consider when refactoring services: responses to services calls should probably have `is_running=False` when `time_left==0`

**Review by benjaminwp18 — CHANGES_REQUESTED** [review]
Move `auxiliary_nodes` and `event_nodes` into a larger `nodes` (`gui_nodes`?) folder with auxiliary and event subfolders

File: src/surface/gui/gui/event_nodes [hunk 3]
```diff
@@ -0,0 +1 @@
+/home/noah/mate/rov-24/src/surface/gui/gui/event_nodes
```
Comment (InvincibleRMC) [inline src/surface/gui/gui/event_nodes:1]: ?

**Review by InvincibleRMC — CHANGES_REQUESTED** [review]

File: src/surface/gui/gui/gui_nodes/auxiliary_nodes/timer.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+from rclpy.executors import MultiThreadedExecutor
+from rclpy.qos import qos_profile_system_default
+from rclpy.time import Time
+from rclpy.duration import Duration
+from rov_msgs.msg import MissionTimerTick
+from rov_msgs.srv import MissionTimerSet
+
+
+PUBLISH_RATE = 10  # Hz
+DEFAULT_DURATION = 15 * 60  # Seconds
```
Comment (InvincibleRMC) [inline src/surface/gui/gui/gui_nodes/auxiliary_nodes/timer.py:13]: Should this be a Duration object?

File: src/surface/gui/gui/gui_nodes/auxiliary_nodes/timer.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        self.set_time_service = self.create_service(
+            MissionTimerSet, "set_mission_timer", callback=self.set_time_callback
+        )
+
+        self.publisher_timer = self.create_timer(1 / PUBLISH_RATE, self.timer_callback)
+
+        self.time_left = Duration(seconds=DEFAULT_DURATION)
+        self.is_running = False
+
+        self.lastTimestamp: Optional[Time] = None
```
Comment (InvincibleRMC) [inline src/surface/gui/gui/gui_nodes/auxiliary_nodes/timer.py:33]: snake_case

File: src/surface/gui/gui/gui_nodes/auxiliary_nodes/timer.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    def do_tick(self) -> None:
+        timestamp = self.get_clock().now()
+
+        if self.is_running and self.lastTimestamp is not None:
+            self.time_left = self.lastTimestamp + self.time_left - timestamp
+
+            if self.time_left < Duration(seconds=0):
```
Comment (InvincibleRMC) [inline src/surface/gui/gui/gui_nodes/auxiliary_nodes/timer.py:41]: Constant for Duration(0)?

File: src/surface/gui/gui/widgets/timer.py [hunk 7]
```diff
@@ -1,72 +1,122 @@
...
+            self.toggle_btn.setText("Start")
 
-    def start_timer(self) -> None:
-        if self.seconds_left == 0:
-            return
-        self.running = True
-        self.timer.start(1000)
-        self.toggle_btn.setText("Pause")
+    def mission_timer_callback(self, msg: MissionTimerTick) -> None:
+        super().mission_timer_callback(msg)
```
Comment (InvincibleRMC) [inline src/surface/gui/gui/widgets/timer.py:98]: I don't think explicit supers should be used unless needed.

**Review by InvincibleRMC — CHANGES_REQUESTED** [review]

File: src/surface/gui/gui/widgets/timer.py [hunk 8]
```diff
@@ -1,72 +1,122 @@
...
+        self.timer_subscription = GUIEventSubscriber(
+            MissionTimerTick,
+            'mission_timer',
+            self.mission_timer_signal,
+        )
+        self.mission_timer_signal.connect(self.mission_timer_callback)
+
         self.running = False
 
         self.label = QLabel('Label')
```
Comment (benjaminwp18) [inline src/surface/gui/gui/widgets/timer.py:30]: This gets used in TimerBase's subclasses. Name it something more better. `timer_display`? `display_label`? `base_display`?

**Review by benjaminwp18 — CHANGES_REQUESTED** [review]
Add docstrings to all methods

**Review by InvincibleRMC — DISMISSED** [review]

File: src/surface/gui/gui/event_nodes (hunk 3 above)
Comment (NoahMollerstuen) [reply src/surface/gui/gui/event_nodes:1]: Idk, this is a normal directory on my machine

File: src/surface/gui/gui/gui_nodes/auxiliary_nodes/timer.py (hunk 6 above)
Comment (NoahMollerstuen) [reply src/surface/gui/gui/gui_nodes/auxiliary_nodes/timer.py:41]: Eh, idk if that helps readability at all. This makes it very clear it's a Duration object. I usually count 0 as an acceptable "magic number"

**Comment by NoahMollerstuen** [comment]
Currently failing the linter because the linter is wrong

**Review by benjaminwp18 — APPROVED** [review]

**Pull request merged**